<a href="https://colab.research.google.com/github/EstibenMT/machine_learning/blob/main/E3/Taller3_Ejercicio_Practico_2_Prediccion_precio_automoviles.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Integrantes del equipo

Grupo: Jueves - Viernes 6:00 a 8:00 pm

- Estiben Montoya Taborda
- Luis David Lopera Parra
- Santiago Gomez Gallego
- Jose Miguel Buritica Morales

## Referencias

- Detrano, R., Janosi, A., Steinbrunn, W., Pfisterer, M., Schmid, J., Sandhu, S., Guppy, K., Lee, S., & Froelicher, V. (1989). *Heart Disease* [Dataset]. UCI Machine Learning Repository. https://doi.org/10.24432/C52P4X
- Schlimmer, J. (1985). *Automobile* [Dataset]. UCI Machine Learning Repository. https://doi.org/10.24432/C5B01C


# Ejercicio Práctico 2: Predicción del precio de automóviles

## Contexto de negocio

A partir de especificaciones técnicas, dimensiones y atributos de distintos vehículos, se busca **predecir el precio de venta del automóvil (`price`)**. Como la variable objetivo es un valor numérico continuo (dinero) y no una categoría, este es un problema de **regresión**, a diferencia del Ejercicio 1 (Heart Disease) que era de clasificación.

El modelo seleccionado será **Linear Regression**, ya que el objetivo es predecir una cantidad numérica continua, siguiendo el mismo criterio de selección usado en el ejercicio de precios de vivienda (Ames Housing) visto en clase.

## Descripción del dataset (UCI Automobile)

El dataset **Automobile** (UCI ML Repository, id=10) contiene 205 registros. UCI define oficialmente `symboling` (riesgo de seguro) como la variable `Target`, y el resto de las 25 columnas —incluida `price`— como `Feature`. Sin embargo, el objetivo de este ejercicio (según la guía del taller) es predecir el **precio** del vehículo, así que más adelante construiremos manualmente nuestra propia variable objetivo a partir de `price`, que llega dentro de `X`.

### Descripción de las variables seleccionadas

| Variable | Nombre | Tipo de dato (dtype) | Rol en UCI | Descripción |
|---|---|---|---|---|
| Riesgo de seguro | `symboling` | `int64` | Target (UCI) | Grado de riesgo asignado al vehículo (-3 = seguro, 3 = riesgoso). No se usa en este ejercicio. |
| Pérdidas normalizadas | `normalized-losses` | `float64` | Feature | Pérdida promedio relativa por tipo de vehículo (tiene valores faltantes). |
| Marca | `make` | `object` | Feature | Fabricante del vehículo (ej. audi, bmw, toyota). |
| Tipo de combustible | `fuel-type` | `object` | Feature | Tipo de combustible (gas, diesel). |
| Aspiración | `aspiration` | `object` | Feature | Tipo de aspiración del motor (std, turbo). |
| Número de puertas | `num-of-doors` | `object` | Feature | Cantidad de puertas escrita en texto (two, four); tiene valores faltantes. |
| Estilo de carrocería | `body-style` | `object` | Feature | Tipo de carrocería (sedan, hatchback, wagon, etc.). |
| Tracción | `drive-wheels` | `object` | Feature | Tipo de tracción (4wd, fwd, rwd). |
| Ubicación del motor | `engine-location` | `object` | Feature | Ubicación del motor (front, rear). |
| Distancia entre ejes | `wheel-base` | `float64` | Feature | Distancia entre ejes delantero y trasero. |
| Longitud | `length` | `float64` | Feature | Longitud total del vehículo. |
| Ancho | `width` | `float64` | Feature | Ancho total del vehículo. |
| Alto | `height` | `float64` | Feature | Altura total del vehículo. |
| Peso en vacío | `curb-weight` | `int64` | Feature | Peso del vehículo sin pasajeros ni carga. |
| Tipo de motor | `engine-type` | `object` | Feature | Tipo de motor (ohc, dohc, l, etc.). |
| Número de cilindros | `num-of-cylinders` | `object` | Feature | Cantidad de cilindros escrita en texto (four, six, etc.). |
| Tamaño del motor | `engine-size` | `int64` | Feature | Cilindraje del motor. |
| Sistema de combustible | `fuel-system` | `object` | Feature | Tipo de sistema de alimentación de combustible. |
| Diámetro del cilindro | `bore` | `float64` | Feature | Diámetro del cilindro del motor (tiene valores faltantes). |
| Carrera del pistón | `stroke` | `float64` | Feature | Recorrido del pistón (tiene valores faltantes). |
| Relación de compresión | `compression-ratio` | `float64` | Feature | Relación de compresión del motor. |
| Potencia | `horsepower` | `float64` | Feature | Potencia del motor en caballos de fuerza (tiene valores faltantes). |
| RPM máximas | `peak-rpm` | `float64` | Feature | Revoluciones por minuto máximas (tiene valores faltantes). |
| Consumo en ciudad | `city-mpg` | `int64` | Feature | Millas por galón en ciudad. |
| Consumo en autopista | `highway-mpg` | `int64` | Feature | Millas por galón en autopista. |
| Precio | `price` | `float64` | Feature (UCI) → **Target (nuestro)** | Precio de venta del vehículo en USD (tiene valores faltantes). Aunque UCI la clasifica como `Feature`, en este ejercicio la usamos como variable objetivo ($y$), tal como pide la guía del taller. |

## Configuración e instalación de `ucimlrepo`

Instalamos la librería oficial de UCI para descargar el dataset directamente desde el repositorio, tal como se indica en la página del dataset.

In [ ]:
!pip install ucimlrepo

from ucimlrepo import fetch_ucirepo

In [ ]:
# fetch dataset
automobile = fetch_ucirepo(id=10)

# data (as pandas dataframes)
X = automobile.data.features
y = automobile.data.targets

# metadata
print(automobile.metadata)

# variable information
print(automobile.variables)

{'uci_id': 10, 'name': 'Automobile', 'repository_url': 'https://archive.ics.uci.edu/dataset/10/automobile', 'data_url': 'https://archive.ics.uci.edu/static/public/10/data.csv', 'abstract': "From 1985 Ward's Automotive Yearbook", 'area': 'Other', 'tasks': ['Regression'], 'characteristics': ['Multivariate'], 'num_instances': 205, 'num_features': 25, 'feature_types': ['Categorical', 'Integer', 'Real'], 'demographics': [], 'target_col': ['symboling'], 'index_col': None, 'has_missing_values': 'yes', 'missing_values_symbol': 'NaN', 'year_of_dataset_creation': 1985, 'last_updated': 'Thu Aug 10 2023', 'dataset_doi': '10.24432/C5B01C', 'creators': ['Jeffrey Schlimmer'], 'intro_paper': None, 'additional_info': {'summary': 'This data set consists of three types of entities: (a) the specification of an auto in terms of various characteristics, (b) its assigned insurance risk rating, (c) its normalized losses in use as compared to other cars.  The second rating corresponds to the degree to which th

Revisamos la salida anterior para conocer la estructura del dataset. Aquí confirmamos algo importante: `automobile.variables` muestra que el rol (`role`) de `price` es `Feature`, y el único `Target` oficial es `symboling`. Como nuestro objetivo (según la guía del taller) es predecir el precio, tendremos que construir nuestra propia variable objetivo a partir de `price`, en vez de usar directamente `y = automobile.data.targets`.

## Preparación de los datos separados

Aquí **no podemos usar `y` directamente**, porque `y = automobile.data.targets` trae `symboling` (el target oficial de UCI), y no `price`. Por eso trabajaremos únicamente sobre `X` (que sí contiene las 25 columnas, incluyendo `price`), y más adelante extraeremos `price` manualmente para construir nuestra propia variable objetivo.

Trabajamos sobre una copia (`features_df`) para no modificar los datos originales devueltos por `fetch_ucirepo`.

In [ ]:
import pandas as pd

features_df = X.copy()

features_df.head()

,price,highway-mpg,city-mpg,peak-rpm,horsepower,compression-ratio,stroke,bore,fuel-system,engine-size,...,length,wheel-base,engine-location,drive-wheels,body-style,num-of-doors,aspiration,fuel-type,make,normalized-losses
0,13495.0,27,21,5000.0,111.0,9.0,2.68,3.47,mpfi,130,...,168.8,88.6,front,rwd,convertible,2.0,std,gas,alfa-romero,NaN
1,16500.0,27,21,5000.0,111.0,9.0,2.68,3.47,mpfi,130,...,168.8,88.6,front,rwd,convertible,2.0,std,gas,alfa-romero,NaN
2,16500.0,26,19,5000.0,154.0,9.0,3.47,2.68,mpfi,152,...,171.2,94.5,front,rwd,hatchback,2.0,std,gas,alfa-romero,NaN
3,13950.0,30,24,5500.0,102.0,10.0,3.40,3.19,mpfi,109,...,176.6,99.8,front,fwd,sedan,4.0,std,gas,audi,164.0
4,17450.0,22,18,5500.0,115.0,8.0,3.40,3.19,mpfi,136,...,176.6,99.4,front,4wd,sedan,4.0,std,gas,audi,164.0


In [ ]:
# y = automobile.data.targets trae 'symboling',
# no es 'price'. No lo usamos porque nuestro objetivo real es predecir el precio.
y.head()

,symboling
0,3
1,3
2,1
3,2
4,2


### Revisión de las variables de entrada

Revisamos la estructura de `features_df` para conocer cuántos registros tiene, qué tipo de datos maneja cada columna y si existen valores faltantes.

In [ ]:
features_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 25 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   price              201 non-null    float64
 1   highway-mpg        205 non-null    int64  
 2   city-mpg           205 non-null    int64  
 3   peak-rpm           203 non-null    float64
 4   horsepower         203 non-null    float64
 5   compression-ratio  205 non-null    float64
 6   stroke             201 non-null    float64
 7   bore               201 non-null    float64
 8   fuel-system        205 non-null    object 
 9   engine-size        205 non-null    int64  
 10  num-of-cylinders   205 non-null    int64  
 11  engine-type        205 non-null    object 
 12  curb-weight        205 non-null    int64  
 13  height             205 non-null    float64
 14  width              205 non-null    float64
 15  length             205 non-null    float64
 16  wheel-base         205 non

In [ ]:
features_df.isna().sum()

,0
price,4
highway-mpg,0
city-mpg,0
peak-rpm,2
horsepower,2
compression-ratio,0
stroke,4
bore,4
fuel-system,0
engine-size,0


In [ ]:
features_df.duplicated().sum()

np.int64(0)

### Revisión de la variable objetivo (`price`)

Como `price` viene dentro de `features_df` (no en un `target_df` separado), la revisamos directamente ahí. Es importante confirmar si tiene valores faltantes, porque una variable objetivo debe estar completa antes de entrenar el modelo.

In [ ]:
features_df['price'].describe()

,price
count,201.000000
mean,13207.129353
std,7947.066342
min,5118.000000
25%,7775.000000
50%,10295.000000
75%,16500.000000
max,45400.000000


In [ ]:
features_df['price'].isna().sum()

np.int64(4)

### Eliminación de registros sin variable objetivo

A diferencia de las demás variables de entrada (que se imputan dentro del pipeline para evitar fuga de datos), los registros que **no tienen valor de `price`** no sirven para entrenar ni para evaluar el modelo, porque no hay nada contra qué comparar la predicción. Por eso, estos registros se eliminan antes de la partición, no se imputan.

In [ ]:
registros_validos = features_df['price'].notna()

features_df_completo = features_df[registros_validos].reset_index(drop=True)

features_df_completo.shape

(201, 25)

## Construcción de la matriz de características (X_model) y el vector objetivo (y_model)

Ahora sí construimos nuestra propia variable objetivo: `y_model` es la columna `price`, y `X_model` son **todas las demás columnas**, eliminando `price` de las features.

In [ ]:
y_model = features_df_completo['price'].copy()
X_model = features_df_completo.drop(columns=['price']).copy()

X_model

,highway-mpg,city-mpg,peak-rpm,horsepower,compression-ratio,stroke,bore,fuel-system,engine-size,num-of-cylinders,...,length,wheel-base,engine-location,drive-wheels,body-style,num-of-doors,aspiration,fuel-type,make,normalized-losses
0,27,21,5000.0,111.0,9.0,2.68,3.47,mpfi,130,4,...,168.8,88.6,front,rwd,convertible,2.0,std,gas,alfa-romero,NaN
1,27,21,5000.0,111.0,9.0,2.68,3.47,mpfi,130,4,...,168.8,88.6,front,rwd,convertible,2.0,std,gas,alfa-romero,NaN
2,26,19,5000.0,154.0,9.0,3.47,2.68,mpfi,152,6,...,171.2,94.5,front,rwd,hatchback,2.0,std,gas,alfa-romero,NaN
3,30,24,5500.0,102.0,10.0,3.40,3.19,mpfi,109,4,...,176.6,99.8,front,fwd,sedan,4.0,std,gas,audi,164.0
4,22,18,5500.0,115.0,8.0,3.40,3.19,mpfi,136,5,...,176.6,99.4,front,4wd,sedan,4.0,std,gas,audi,164.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
196,28,23,5400.0,114.0,9.5,3.15,3.78,mpfi,141,4,...,188.8,109.1,front,rwd,sedan,4.0,std,gas,volvo,95.0
197,25,19,5300.0,160.0,8.7,3.15,3.78,mpfi,141,4,...,188.8,109.1,front,rwd,sedan,4.0,turbo,gas,volvo,95.0
198,23,18,5500.0,134.0,8.8,2.87,3.58,mpfi,173,6,...,188.8,109.1,front,rwd,sedan,4.0,std,gas,volvo,95.0
199,27,26,4800.0,106.0,23.0,3.40,3.01,idi,145,6,...,188.8,109.1,front,rwd,sedan,4.0,turbo,diesel,volvo,95.0


In [ ]:
y_model

,price
0,13495.0
1,16500.0
2,16500.0
3,13950.0
4,17450.0
...,...
196,16845.0
197,19045.0
198,21485.0
199,22470.0


## Partición en conjuntos de entrenamiento y prueba

Dividimos los datos en un conjunto de entrenamiento (80%) y uno de prueba (20%). A diferencia del Ejercicio 1 (clasificación), aquí **no se usa `stratify`**, porque ese parámetro solo aplica quando la variable objetivo es categórica (para mantener la proporción de clases); como `price` es una variable numérica continua, no tiene sentido estratificar por ella.

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_model, y_model, test_size=0.2, random_state=917)

## Verificación del tipo de dato de cada columna

Antes de construir el pipeline de preprocesamiento, verificamos con `dtypes` qué tipo de dato tiene cada columna. **las columnas de tipo `object` (texto) van al pipeline categórico**; **las columnas numéricas (`int64`/`float64`) van al pipeline numérico**.

In [ ]:
X_model.dtypes

,0
highway-mpg,int64
city-mpg,int64
peak-rpm,float64
horsepower,float64
compression-ratio,float64
stroke,float64
bore,float64
fuel-system,object
engine-size,int64
num-of-cylinders,int64


### Identificación automática de columnas numéricas y categóricas

Como este dataset tiene 24 columnas de entrada en `X_model` (25 originales menos `price`, que ahora es nuestro target), aquí conviene usar la técnica automática con `select_dtypes`.

In [ ]:
columnas_numericas = X_model.select_dtypes(include=['int64', 'float64']).columns
columnas_numericas

Index(['highway-mpg', 'city-mpg', 'peak-rpm', 'horsepower',
       'compression-ratio', 'stroke', 'bore', 'engine-size',
       'num-of-cylinders', 'curb-weight', 'height', 'width', 'length',
       'wheel-base', 'num-of-doors', 'normalized-losses'],
      dtype='object')

In [ ]:
columnas_categoricas = X_model.select_dtypes(include=['object']).columns
columnas_categoricas

Index(['fuel-system', 'engine-type', 'engine-location', 'drive-wheels',
       'body-style', 'aspiration', 'fuel-type', 'make'],
      dtype='object')

## Definición de transformaciones para variables numéricas y categóricas

Para las variables numéricas se aplica imputación por mediana (varias columnas como `normalized-losses`, `bore`, `stroke`, `horsepower` o `peak-rpm` tienen valores faltantes, y la mediana es robusta a valores atípicos) seguida de escalado estándar, necesario para que el modelo trate todas las variables en una escala comparable.

Para las variables categóricas se aplica imputación por el valor más frecuente (por ejemplo `num-of-doors` tiene un par de valores faltantes) seguida de codificación One-Hot, eliminando la primera categoría para evitar colinealidad.

In [ ]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler

pipeline_numerico = Pipeline(steps=[('imputacion_numerica', SimpleImputer(strategy='median')), ('escalado_numerico', StandardScaler())])

In [ ]:
from sklearn.preprocessing import OneHotEncoder

pipeline_categorico = Pipeline(steps=[('imputacion_categorica', SimpleImputer(strategy='most_frequent')), ('vdummies', OneHotEncoder(drop='first', handle_unknown='ignore'))])

## Integración de los transformadores en un ColumnTransformer

Despues de la separación es necesario combinar ambos pipelines, porque el dataset tiene tanto columnas `object` (categóricas) como numéricas.

In [ ]:
from sklearn.compose import ColumnTransformer

preprocesador = ColumnTransformer(transformers=[('num', pipeline_numerico, list(columnas_numericas)), ('cat', pipeline_categorico, list(columnas_categoricas))])

## Construcción y entrenamiento del pipeline final con Linear Regression

El modelo seleccionado es **Linear Regression**, ya que el objetivo es predecir un valor numérico continuo (el precio del vehículo).

In [ ]:
from sklearn.linear_model import LinearRegression

automobile_pipeline = Pipeline(steps=[('preprocesamiento', preprocesador), ('regresor', LinearRegression())])

automobile_pipeline.fit(X_train, y_train)

Pipeline(steps=[('preprocesamiento',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('imputacion_numerica',
                                                                   SimpleImputer(strategy='median')),
                                                                  ('escalado_numerico',
                                                                   StandardScaler())]),
                                                  ['highway-mpg', 'city-mpg',
                                                   'peak-rpm', 'horsepower',
                                                   'compression-ratio',
                                                   'stroke', 'bore',
                                                   'engine-size',
                                                   'num-of-cylinders',
                                                   'curb-weight', 'height',
                                                   'width', 'length',
                                                   'wheel-base', 'num-of-doors',
                                                   'normalized-losses']),
                                                 ('cat',
                                                  Pipeline(steps=[('imputacion_categorica',
                                                                   SimpleImputer(strategy='most_frequent')),
                                                                  ('vdummies',
                                                                   OneHotEncoder(drop='first',
                                                                                 handle_unknown='ignore'))]),
                                                  ['fuel-system', 'engine-type',
                                                   'engine-location',
                                                   'drive-wheels', 'body-style',
                                                   'aspiration', 'fuel-type',
                                                   'make'])])),
                ('regresor', LinearRegression())])

## Predicción y evaluación del modelo

Generamos las predicciones sobre el conjunto de prueba y calculamos el **RMSE** (Root Mean Squared Error), la métrica estándar para evaluar modelos de regresión.

In [ ]:
y_pred = automobile_pipeline.predict(X_test)

y_pred

/usr/local/lib/python3.13/dist-packages/sklearn/preprocessing/_encoders.py:246: UserWarning: Found unknown categories in columns [0] during transform. These unknown categories will be encoded as all zeros
  warnings.warn(


array([10817.94288611, 16264.3942131 , 16025.71116462, 16869.2281678 ,
       10888.96874198,  7735.30724893,  9323.34729992,  9095.41118871,
       36269.50272077,  6888.34462093,  8554.58537852,  9905.05060184,
        6058.73254456,  6852.22464083,  9556.96170412, 20530.67592371,
       18282.73502499, 10619.93492799, 10175.36742524,  7938.3580904 ,
        6566.41740917, 38251.92106965, 13998.92206793,  5976.92891981,
        8236.62250048,  6978.35880285, 18905.14762017, 11718.97720648,
       10875.58178651,  6032.0885452 , 32459.8453586 , 13358.39810802,
       10610.43962194, 13971.94081512, 26504.87227056, 22609.70464239,
       10875.58178651,  6115.11225934,  6369.90458561, 18427.34581048,
        7551.84045612])

In [ ]:
from sklearn.metrics import root_mean_squared_error

rmse = root_mean_squared_error(y_test, y_pred)
rmse

2631.019750119268

### Interpretación del resultado

El modelo de regresión lineal obtuvo un **RMSE de 2631,02 dólares**.

Esto significa que las predicciones del precio se alejan del valor real aproximadamente en **2631 dólares**, aunque el error puede ser mayor o menor dependiendo del vehículo.

Para interpretar mejor este resultado, lo comparamos con los precios del conjunto de datos:

- Precio promedio: **13 207,13 dólares**.
- Precio mínimo: **5118 dólares**.
- Precio máximo: **45 400 dólares**.
- Desviación estándar: **7947,07 dólares**.

El RMSE representa aproximadamente el **19,9 % del precio promedio**. Esto indica que el modelo logra realizar predicciones razonables, pero todavía presenta diferencias importantes en algunos vehículos, especialmente cuando sus precios son muy altos o tienen características poco comunes.

En conclusión, el modelo aprendió una relación útil entre las características técnicas del automóvil y su precio. Sin embargo, el error de 2631 dólares muestra que la regresión lineal no logra explicar completamente todas las variaciones del precio.

In [ ]:
y_model.describe()

,price
count,201.000000
mean,13207.129353
std,7947.066342
min,5118.000000
25%,7775.000000
50%,10295.000000
75%,16500.000000
max,45400.000000


## Declaración sobre uso de IA Generativa

Se utilizó una herramienta de IA Generativa (Claude) como apoyo para:

- Redactar las celdas de Markdown explicativas y su formato.
- Consultar sobre el detalle de los métodos utilizados y más información de las librerías utilizadas.
- Explicación de la construcción de los pipelines (`pipeline_numerico`, `pipeline_categorico`).

**Prompts utilizados (resumen):**

1. "Ayúdame a redactar las celdas de Markdown explicando cada paso del flujo de trabajo que te voy pasando y el detalle con el que te oriento."
2. "Explícame cómo funciona `select_dtypes` en pandas para identificar columnas numéricas y categóricas."
3. "Explícame por qué no se usa `stratify` en `train_test_split` cuando el problema es de regresión."
4. "Explícame qué hace cada paso de `pipeline_numerico` y `pipeline_categorico` y por qué es necesario según los datos que tengo."

Todo el código fue revisado, ejecutado y validado por el equipo antes de la entrega, y las interpretaciones de los resultados fueron completadas por el equipo con base en los valores obtenidos.
